src/common-libs/plan/index.ts

```diff
@@ -0,0 +1,511 @@
+import { Database } from '@unocha/hpc-api-core/src/db';
+import { PlanReportingPeriodId } from '@unocha/hpc-api-core/src/db/models/planReportingPeriod';
+import { InstanceDataOfModel } from '@unocha/hpc-api-core/src/db/util/raw-model';
+import { Brand, createBrandedValue } from '@unocha/hpc-api-core/src/util/types';
+import { AddPlanTagInput } from '../../domain-services/plantag/graphql/types';
+
+export const getVersion = async (
+  models: Database,
+  planTag: AddPlanTagInput
+): Promise<string> => {
+  const latestPlanTag = await models.planTag.find({
+    where: {
+      planId: createBrandedValue(planTag.planId),
+    },
+    orderBy: { column: 'createdAt', order: 'desc' },
+  });
+  const latestTagName =
+    (latestPlanTag.length && latestPlanTag[0].name) || '0.0';
+  let [majorVersion, minorVersion] = latestTagName
+    .split('.')
+    .map((v) => parseInt(v) || 0);
+  if (planTag.major) {
+    majorVersion++;
+    minorVersion = 0;
+  } else {
+    minorVersion++;
+  }
+  return `${majorVersion}.${minorVersion}`;
+};
+
+export const setPlanReportingPeriod = async (
+  models: Database,
+  planTag: AddPlanTagInput
+) => {
+  const reportingPeriods = await models.planReportingPeriod.find({
+    where: {
+      planId: createBrandedValue(planTag.planId),
+    },
+    orderBy: { column: 'periodNumber', order: 'asc' },
+  });
+  if (!reportingPeriods.length) {
+    throw new Error(
+      'Cannot activate planTag if associated plan has no reporting period'
+    );
+  }
+  const latestPlanVersion = await models.planVersion.findOne({
+    where: {
+      planId: createBrandedValue(planTag.planId),
+      latestVersion: true,
+    },
+  });
+  if (latestPlanVersion && !latestPlanVersion.currentReportingPeriodId) {
+    await models.planVersion.update({
+      values: {
+        currentReportingPeriodId: reportingPeriods[0].id,
+      },
+      where: {
+        id: latestPlanVersion.id,
+      },
+    });
+  }
+};
+
+const getAssociatedPlanVersion = async (
+  models: Database,
+  planTag: InstanceDataOfModel<Database['planTag']>
+) => {
+  const associatedPlanVersion = await models.planVersion.findOne({
+    where: {
+      planId: planTag.planId,
+      createdAt: {
+        [models.Op.LTE]: planTag.createdAt,
+      },
+      latestVersion: true,
+    },
+  });
+  if (!associatedPlanVersion) {
+    throw new Error('Cannot activate planTag if no associated version exist');
+  }
+  return associatedPlanVersion;
+};
+
+export const updateLastPublishedReportingPeriodId = async (
+  models: Database,
+  planTag: InstanceDataOfModel<Database['planTag']>
+) => {
+  const reportingPeriods = await models.planReportingPeriod.find({
+    where: {
+      planId: planTag.planId,
+      measurementsGenerated: true,
+    },
+    orderBy: { column: 'periodNumber', order: 'desc' },
+  });
+
+  if (reportingPeriods.length) {
```

Comment:
You can reduce indentation here
```suggestion
  if (!reportingPeriods.length) {
    return;
  }
```